src/lib.rs

```diff
@@ -1,12 +1,253 @@
 //! Sync tool between other sources and our infrastructure based on Zitadel.
+use anyhow::{Context, Result};
+use futures::{Stream, StreamExt};
+use user::{StringOrBytes, User};
+use zitadel::Zitadel;
 
 mod config;
 mod sources;
 mod user;
 mod zitadel;
 
+use std::collections::VecDeque;
+
 pub use config::{Config, FeatureFlag};
 pub use sources::{
 	csv::test_helpers as csv_test_helpers, ldap::AttributeMapping,
 	ukt::test_helpers as ukt_test_helpers,
 };
+use sources::{csv::CsvSource, ldap::LdapSource, ukt::UktSource, Source};
+
+/// Helper function to add metadata to streamed zitadel users
+// TODO: If async closures become a reality, this should be factored
+// into the `zitadel::search_result_to_user` function
+async fn get_next_zitadel_user(
+	stream: &mut (impl Stream<Item = Result<(User, String)>> + Send + Unpin),
+	zitadel: &mut Zitadel,
+) -> Result<Option<(User, String)>> {
+	match stream.next().await.transpose()? {
+		Some(mut zitadel_user) => {
+			let preferred_username: Option<StringOrBytes> = zitadel
+				.zitadel_client
+				.get_user_metadata(&zitadel_user.1, "preferred_username")
+				.await
+				.ok()
+				.and_then(|metadata| metadata.metadata().value())
+				.map(Into::into);
+
+			zitadel_user.0.preferred_username = preferred_username;
+
+			Ok(Some(zitadel_user))
+		}
+		None => Ok(None),
+	}
+}
+
+/// Perform a sync operation
+pub async fn perform_sync(config: &Config) -> Result<()> {
+	/// Get users from a source
+	async fn get_users_from_source(source: impl Source + Send) -> Result<VecDeque<User>> {
+		source
+			.get_sorted_users()
+			.await
+			.map(VecDeque::from)
+			.context(format!("Failed to query users from {}", source.get_name()))
+	}
+
+	let csv = config.sources.csv.clone().map(CsvSource::new);
+	let ldap = config.sources.ldap.clone().map(LdapSource::new);
+	let ukt = config.sources.ukt.clone().map(UktSource::new);
+
+	// The ukt source is handled specially, since it doesn't behave as
+	// the others
+	if let Some(ukt) = ukt {
```

Comment:
This doesn't correctly check that exactly *one* source is defined, which is unfortunate. Couldn't think of a good way to express this condition, while also treating ukt specially, but DRY-ing out the other sources.